server/sonar-server/src/main/java/org/sonar/server/test/ws/TestsListAction.java

```diff
@@ -0,0 +1,221 @@
+      request.mandatoryParamAsInt(WebService.Param.PAGE),
+      request.mandatoryParamAsInt(WebService.Param.PAGE_SIZE)
+      );
+
+    checkArguments(testUuid, testFileUuid, testFileKey, sourceFileUuid, sourceFileLineNumber);
+    checkPermissions(testUuid, testFileUuid, testFileKey, sourceFileUuid);
+    SearchResult<TestDoc> tests = searchTests(testUuid, testFileUuid, testFileKey, sourceFileUuid, sourceFileLineNumber, searchOptions);
+    Map<String, ComponentDto> componentsByTestFileUuid = buildComponentsByTestFileUuid(tests.getDocs());
+
+    JsonWriter json = response.newJsonWriter().beginObject();
+    writeTests(tests.getDocs(), componentsByTestFileUuid, json);
+    searchOptions.writeJson(json, tests.getTotal());
+    json.endObject().close();
+  }
+
+  private void writeTests(List<TestDoc> tests, Map<String, ComponentDto> componentsByTestFileUuid, JsonWriter json) {
+    json.name("tests").beginArray();
+    for (TestDoc test : tests) {
+      String fileUuid = test.fileUuid();
+      json.beginObject();
+      json.prop("testUuid", test.testUuid());
+      json.prop("fileUuid", fileUuid);
+      json.prop("name", test.name());
+      json.prop("status", test.status());
+      json.prop("durationInMs", test.durationInMs());
+      json.prop("message", test.message());
+      json.prop("stacktrace", test.stackTrace());
+      json.prop("coveredLines", coveredLines(test.coveredFiles()));
+      json.prop("fileKey", componentsByTestFileUuid.get(fileUuid).key());
+      json.prop("fileLongName", componentsByTestFileUuid.get(fileUuid).longName());
+      json.endObject();
+    }
+    json.endArray();
+  }
+
+  private long coveredLines(List<CoveredFileDoc> coveredFiles) {
+    long numberOfLinesCovered = 0L;
+    for (CoveredFileDoc coveredFile : coveredFiles) {
+      numberOfLinesCovered += coveredFile.coveredLines().size();
+    }
+
+    return numberOfLinesCovered;
+  }
+
+  private Map<String, ComponentDto> buildComponentsByTestFileUuid(List<TestDoc> tests) {
+    List<String> fileUuids = Lists.transform(tests, new NonNullInputFunction<TestDoc, String>() {
+      @Override
+      protected String doApply(TestDoc testDoc) {
+        return testDoc.fileUuid();
+      }
+    });
+    DbSession dbSession = dbClient.openSession(false);
+    List<ComponentDto> components;
+    try {
+      components = dbClient.componentDao().getByUuids(dbSession, fileUuids);
+    } finally {
+      MyBatis.closeQuietly(dbSession);
+    }
+
+    return Maps.uniqueIndex(components, new Function<ComponentDto, String>() {
+      @Override
+      public String apply(@Nullable ComponentDto componentDto) {
+        return componentDto.uuid();
+      }
+    });
+  }
+
+  private SearchResult<TestDoc> searchTests(@Nullable String testUuid, @Nullable String testFileUuid, @Nullable String testFileKey, @Nullable String sourceFileUuid,
+    @Nullable Integer sourceFileLineNumber,
+    SearchOptions searchOptions) {
+    if (testUuid != null) {
```

Comment:
Instead of having all these checks in each method to know in which "state" you are (using testUuid or using  testFileUuid or using sourceFileUuid and sourceFileLineNumber), don't you think it would have been better to have 3 methods for each use case that : 
- checkPermissions
- execute the search

I think it would improve the readability of this class, as it will remove all this nullable parameters in each methods